modutils/man/ksyms.1
.\" Copyright (c) 1996 Free Software Foundation, Inc.
.\" This program is distributed according to the Gnu General Public License.
.\" See the file COPYING in the kernel source directory
.\" $Id: ksyms.1,v 1.1.1.1 1998/01/06 20:51:07 ewt Exp $
.\"
.TH KSYMS 1 "26 Dec 1996" Linux "Linux Module Support"
.SH NAME
ksyms \- display exported kernel symbols.
.SH SYNOPSIS
.B ksyms
[ \-a ] [ \-h ] [ \-m ]
.SH DESCRIPTION
.B Ksyms
shows information about exported kernel symbols.  The format is
address, name, and defining module.
.SS OPTIONS
.TP
.I \-a
Display all symbols.  By default, symbols from the kernel proper
are suppressed.
.TP
.I \-h
Suppress the column header.
.TP
.I \-m
Display module information.  Includes each module's kernel load address
and size in the listing.
.SH SEE ALSO
insmod(1), modprobe(1), depmod(1), rmmod(1), lsmod(1), modules(2)
.SH HISTORY
The
.B ksyms
command was first concieved by Bjorn Ekwall <bj0rn@blox.se>
.br
The '-m' option was inspired by David Hinds <dhinds@allegro.stanford.edu>
.br
Updated for 2.1.17 by Richard Henderson <rth@tamu.edu>


modutils/ChangeLog
Thu Sep 11 08:43:03 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* Release version 2.1.55.

Wed Sep 10 16:34:41 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* genksyms/genksyms.c (crc_prefix): New variable.
	(export_symbol): Use it.
	(main): New option -p sets it.
	* man/genksyms.8: Update documentation.
	Idea from Jacques Gelinas <jack@solucorp.qc.ca>.

Wed Sep 10 16:09:37 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* insmod/Makefile: Allow insmod & rmmod to run as a combined binary.
	* insmod/main.c: New file.
	* Makeconfig (COMBINE_INSMOD_RMMOD): New define.
	Requested by Erik Troan <ewt@redhat.com>.

Wed Sep 10 15:41:16 PDT 1997  Richard Henderson  <rth@cygnus.com>

	* depmod/depmod.c (load_obj_file): Print error if fopen fails.
	(main): Continue processing on errors, so that we see all errors
	all at once.
	Reported by Gunther Mayer <gunther.mayer@braunschweig.netsurf.de>.

Tue Aug 26 11:14:36 -0400 1997  Telly Mavroidis  <mavroidi@ctc-fund.com>

	* util/sys_qm.c [__sparc__]: Fix typo.

Tue Jul 29 23:02:04 PDT 1997  Richard Henderson  <richard@gnu.ai.mit.edu>

	* include/module.h: Rename sizeof_type to tgt_sizeof_type.  Kill
	the struct duplication by defining a tgt_long.
	* insmod/insmod.c: Propogate sizeof_type change.

Tue Jul 29 22:23:16 PDT 1997  Jakub Jelinek  <jj@sunsite.mff.cuni.cz>

	* insmod/insmod.c (search_module_path): Add ipv6.

Tue Jul 29 22:23:16 PDT 1997  Jakub Jelinek  <jj@sunsite.mff.cuni.cz>

	Sparc64 support:
	* depmod/Makefile (DEFS): Pass in ARCH as well.
	* insmod/Makefile: Likewise.
	* obj/Makefile: Likewise.

	* insmod/insmod.c: Don't use native sizeof(type), use constants
	sizeof_type that correspond to the target.
	* include/module.h: Detect and cope with sparc64 target and
	sparc32 host.

	* include/obj.h: libc5 doesn't declare ELF64_ST_INFO either.

	* include/elf_sparc64.h: New file.
	* obj/obj_sparc64.c: New file.

Sat Jul 19 23:56:02 +0200 1997  Florian La Roche  <florian@knorke.saar.de>

	* depmod/misc.c (resolve_string): Kill assignment warning.
	* genksyms/genksyms.h: Extern outfile and debugfile properly.
	* man/modprobe.1: New file.

Wed Apr 16 15:27:17 +1000 1997  Paul Mackerras  <paulus@cs.anu.edu.au>

	* include/elf_ppc.h, obj/obj_ppc.c: New files.

Mon Apr 14 20:33:56 CDT 1997  Richard Henderson  <rth@tamu.edu>

	* depmod/depmod.c (print_deps_file): Fix initial array
	allocation size.

Sat Apr 12 19:23:25 CDT 1997  Richard Henderson  <rth@tamu.edu>

	* obj/obj_reloc.c (obj_relocate): If AXP_BROKEN_GAS, work
	around the old gas local symbol LITERAL reloc bug.

Mon Mar 31 23:03:12 EST 1997  David S. Miller  <davem@jenolan.rutgers.edu>

	* genksyms/lex.l (yylex) [case ST_TABLE_5]: Missing break.

Tue Mar 25 21:41:00 CST 1997  Richard Henderson  <rth@tamu.edu>

	* depmod/depmod.c (old_read_kernel_syms): Get symbols via
	get_kernel_syms syscall rather than /proc/ksyms.

	* depmod/modprobe.c (is_removable): Final arg to query_module
	is a size_t, not an int.

Sat Mar 15 19:34:07 PST 1997  H.J. Lu  <hjl@gnu.ai.mit.edu>

	* depmod/conf_file.c (release_all_sets): Set mod_set to NULL.
	(read_config_file): only call release_all_sets () once.
	call resolve_string () to expand string.

	* depmod/misc.c (resolve_string): New function that resolves
	string with `foo`.
	(what_command): New function that returns the known commands.
	(expand_command): New funtion that expands pattern or
	command. Only `uname -r` is supported this time.

	* depmod/misc.h: New prototype for resolve_string ().

	* depmod/modprobe.c (read_deps_file): call resolve_string () to
	expand string.

	* man/depmod.1: Fix typos. The syntax should be "path[TAG]=PATHNAME"
	means TAG is appended to PATHNAME.

Thu Mar 13 13:38:12 CST 1997  Richard Henderson  <rth@tamu.edu>

	* insmod/insmod.c (ncv_strcmp): New function that combines old
	m_strcmp and k_strcmp -- it's not a simple matter to always have
	the versioned symbol as the first or second argument.
	(main): Install ncv_strcmp instead.
	Reported by David Bourgin <dbourgin@wsc.com>.

	* insmod/insmod.c (get_kernel_version): Take an extra argument to
	return the full uname string.
	(old_get_module_version): Likewise.
	(new_get_module_version): Likewise.
	(main): Record and compare the full uname string, to handle
	suffixes like "-ISS" and "-SMP".

Sun Mar  9 23:52:18 CST 1997  Richard Henderson  <rth@tamu.edu>

	* genksyms/lex.l (yylex): Only use the file component of the input
	path when generating the output filename.
	Reported by Markus Dickebohm <m.dickebohm@uni-koeln.de>.

	* depmod/Makefile (clean): Remove modprobe and depmod.

Sun Mar  9 13:55:27 MET 1997  Eddie C. Dost  <ecd@skynet.be>

	* obj/obj_sparc.c (arch_apply_relocation): Handle lots
	more relocation types.

Tue Mar  4 08:12:58 CST 1997  Richard Henderson  <rth@tamu.edu>

	* Merge new depmod from Marcin Dalecki
	<dalecki@sub994.sub.uni-goettingen.de>.

	* insmod/insmod.c (get_kernel_version): Don't assert that the last
	number of the version be followed by a NUL.
	(old_get_module_version, new_get_module_version): Likewise.
	Reported by Keith Owens <kaos@ocs.com.au>

	* insmod/insmod.c: Merge character and string array patches
	from Jean Tourrilhes <jt@hplb.hpl.hp.com>.

	* kerneld/kerneld.c (kerneld_error): Add missing va_end; use vsnprintf.

Sun Feb  2 13:21:39 CST 1997  Richard Henderson  <rth@tamu.edu>

	* genksyms/lex.l (yylex): Update documentation on 2.0.x symbol table
	processing.  Generate phrases for asm blocks as well, as it is more
	efficient to do it in the lexer.
	* genksyms/parse.y (opt_asm_phrase, asm_declaration): Recognise 
	the ASM_PHRASE token rather than trying to work it out ourselves.

Sun Feb  2 07:44:00 CST 1997  Richard Henderson  <rth@tamu.edu>

	* man/kerneld.8: Format for section 8.
	* man/genksyms.8: Renamed from genksyms.1 to match how it formats.
	Reported by Martin von Loewis <martin@mira.isdn.cs.tu-berlin.de>.

Tue Jan 28 21:10:07 MET 1997  Janos Farkas  <chexum@shadow.banki.hu>


modutils/insmod/ksyms.c
/* Display the symbols exported from the running kernel.
   Copyright 1996, 1997 Linux International.

   New implementation contributed by Richard Henderson <rth@tamu.edu>
   Based on original work by Bjorn Eckwall <bj0rn@blox.se>

   This file is part of the Linux modutils.

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2 of the License, or (at your
   option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#ident "$Id: ksyms.c,v 1.2 1999/09/11 15:43:57 msw Exp $"

#include <sys/types.h>
#include <stdio.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>

#include "module.h"
#include "util.h"
#include "version.h"

#include "logger.h"


/*======================================================================*/

int flag_allsyms = 0;
int flag_mod_info = 0;
int flag_print_header = 1;


/*======================================================================*/

static inline void
print_symbol(const char *name, unsigned long value, const char *module)
{
  if (module)
    printf("%0*lx  %-32s  [%s]\n", (int)(2*sizeof(void*)),
	   value, name, module);
  else
    printf("%0*lx  %-32s\n", (int)(2*sizeof(void*)), value, name);
}

static inline void
print_mod_info(const char *module, unsigned long addr, unsigned long size)
{
  int o;

  o = printf("%0*lx  (%luk)", (int)(2*sizeof(void*)), addr,
	     (size / 1024) + ((size % 1024) != 0));

  o = 2*sizeof(void*)+2+32+2 - o;

  printf("%*s[%s]\n", o, "", module);
}


/* If we don't have query_module()... */

static int
old_ksyms(void)
{
  struct old_kernel_sym *ksyms, *k;
  int nsyms, i;
  int kmem_fd = -1;
  const char *module;

  nsyms = get_kernel_syms(NULL);
  if (nsyms < 0)
    {
      perror("get_kernel_syms");
      return 1;
    }

  ksyms = xmalloc(nsyms * sizeof(*ksyms));

  if (get_kernel_syms(ksyms) != nsyms)
    {
      fprintf(stderr, "Inconsistency reading kernel symbols -- "
	      "is someone else playing with modules?\n");
      return 1;
    }

  /* If requested, open kmem so we can get at module information.  */

  if (flag_mod_info)
    {
      kmem_fd = open("/dev/kmem", O_RDONLY);
      if (kmem_fd < 0)
	{
	  perror("ksyms: open /dev/kmem");
	  return 1;
	}
    }

  module = NULL;
  for (k = ksyms, i = 0; i < nsyms; ++i, ++k)
    if (k->name[0] == '#')
      {
	if (k->name[1])
	  {
	    module = &k->name[1];
	    if (flag_mod_info)
	      {
		struct old_module info;

		if (lseek(kmem_fd, (off_t)k->value, SEEK_SET) == -1
		    || read(kmem_fd, &info, sizeof(info)) != sizeof(info))
		  {
		    perror("ksyms: /dev/kmem");
		    return 1;
		  }
		print_mod_info(module, info.addr, info.size * getpagesize());
	      }
	  }
	else
	  {
	    if (!flag_allsyms)
	      break;
	    module = NULL;
	  }
      }
    else
      print_symbol(k->name, k->value, module);

  free(ksyms);
  if (flag_mod_info)
    close(kmem_fd);

  return 0;
}


/* If we do have query_module()...  */

static int
new_ksyms(void)
{
  char *modules;
  size_t nmodules;
  struct new_module_symbol *syms;
  size_t nsymbols;

  size_t ret;
  size_t bufsize;
  size_t i, j;
  char *m;
  struct new_module_symbol *s;

  /* Query for the module names.  */

  modules = xmalloc(bufsize = 256);
retry_modules_load:
  if (query_module(NULL, QM_MODULES, modules, bufsize, &ret))
    {
      if (errno == ENOSPC)
	{
	  modules = xrealloc(modules, bufsize = ret);
	  goto retry_modules_load;
	}
      perror("ksyms: QM_MODULES");
      return 1;
    }
  nmodules = ret;

  /* Iterate over the modules, in order, and care for their symbols.  */

  syms = xmalloc(bufsize = 16*1024);
  for (i = 0, m = modules; i < nmodules; ++i, m += strlen(m)+1)
    {
      if (flag_mod_info)
	{
	  struct new_module_info info;
	  if (query_module(m, QM_INFO, &info, sizeof(info), &ret))
	    {
	      if (errno == ENOENT)
		{
		  /* The module must have been removed in the interim: skip */
		  continue;
		}
	      perror("ksyms: QM_INFO");
	      return 1;
	    }
	  print_mod_info(m, info.addr, info.size);
	}

    retry_mod_symbol_load:
      if (query_module(m, QM_SYMBOLS, syms, bufsize, &ret))
	switch (errno)
	  {
	  case ENOSPC:
	    syms = xrealloc(syms, bufsize = ret);
	    goto retry_mod_symbol_load;
	  case ENOENT:
	    continue;
	  default:
	    perror("ksyms: QM_SYMBOLS");
	    return 1;
	  }
      nsymbols = ret;

      for (j = 0, s = syms; j < nsymbols; ++j, ++s)
	print_symbol((char *)syms + s->name, s->value, m);
    }

  /* Now do the kernel itself.  */

  if (flag_allsyms)
    {
    retry_kern_symbol_load:
      if (query_module(NULL, QM_SYMBOLS, syms, bufsize, &ret))
	{
	  if (errno == ENOSPC)
	    {
	      syms = xrealloc(syms, bufsize = ret);
	      goto retry_kern_symbol_load;
	    }
	  perror("ksyms: QM_SYMBOLS");
	  return 1;
	}
      nsymbols = ret;

      for (j = 0, s = syms; j < nsymbols; ++j, ++s)
	print_symbol((char *)syms + s->name, s->value, NULL); 
    }

  free(modules);
  free(syms);

  return 0;
}


/*======================================================================*/

int main(int argc, char **argv)
{
  int i;

  error_file = "ksyms";

  while ((i = getopt(argc, argv, "amhV")) != EOF)
    switch (i)
      {
      case 'a':
	flag_allsyms = 1;
	break;
      case 'm':
	flag_mod_info = 1;
	break;
      case 'h':
	flag_print_header = 0;
	break;
      case 'V':
        fputs("ksyms version " MODUTILS_VERSION "\n", stderr);
        break;
      }

  if (flag_print_header)
    printf("%-*s  %-32s  Defined by\n", (int)(2*sizeof(void*)),
	   "Address", "Symbol");

  if (query_module(NULL, 0, NULL, 0, NULL) == 0)
    return new_ksyms();
  else
    return old_ksyms();
}
